contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,474 @@
+  if (!already_send_resp_headers_) {
+    // we can send resp headers only one time.
+    if (!resp_headers_->Status()) {
+      // if go side not set status, c++ side set default status
+      resp_headers_->setStatus(static_cast<uint64_t>(HttpStatusCode::Success));
+    }
+    upstream_request_->decodeHeaders(std::move(resp_headers_), false);
+    already_send_resp_headers_ = true;
+  }
+  upstream_request_->decodeData(data, end_stream);
+}
+
+void TcpUpstream::onEvent(Network::ConnectionEvent event) {
+  if (event != Network::ConnectionEvent::Connected && upstream_request_) {
+    upstream_request_->onResetStream(Envoy::Http::StreamResetReason::ConnectionTermination, "");
+  }
+}
+
+void TcpUpstream::onAboveWriteBufferHighWatermark() {
+  ENVOY_LOG(debug, "tcp upstream onAboveWriteBufferHighWatermark");
+  if (upstream_request_) {
+    upstream_request_->onAboveWriteBufferHighWatermark();
+  }
+}
+
+void TcpUpstream::onBelowWriteBufferLowWatermark() {
+  ENVOY_LOG(debug, "tcp upstream onBelowWriteBufferLowWatermark");
+  if (upstream_request_) {
+    upstream_request_->onBelowWriteBufferLowWatermark();
+  }
+}
+
+void copyHeaderMapToGo(const Envoy::Http::HeaderMap& m, GoString* go_strs, char* go_buf) {
+  auto i = 0;
+  m.iterate([&i, &go_strs, &go_buf](const Envoy::Http::HeaderEntry& header) -> Envoy::Http::HeaderMap::Iterate {
+    // It's safe to use StringView here, since we will copy them into Golang.
+    auto key = header.key().getStringView();
+    auto value = header.value().getStringView();
+
+    auto len = key.length();
+    // go_strs is the heap memory of go, and the length is twice the number of headers. So range it
+    // is safe.
+    go_strs[i].n = len;
+    go_strs[i].p = go_buf;
+    // go_buf is the heap memory of go, and the length is the total length of all keys and values in
+    // the header. So use memcpy is safe.
+    memcpy(go_buf, key.data(), len); // NOLINT(safe-memcpy)
+    go_buf += len;
+    i++;
+
+    len = value.length();
+    go_strs[i].n = len;
+    // go_buf may be an invalid pointer in Golang side when len is 0.
+    if (len > 0) {
+      go_strs[i].p = go_buf;
+      memcpy(go_buf, value.data(), len); // NOLINT(safe-memcpy)
+      go_buf += len;
+    }
+    i++;
+    return Envoy::Http::HeaderMap::Iterate::Continue;
+  });
+}
+
+CAPIStatus TcpUpstream::copyHeaders(ProcessorState& state, GoString* go_strs, char* go_buf) {
+  auto headers = state.headers;
+  if (headers == nullptr) {
+    ENVOY_LOG(debug, "tcp upstream invoking cgo api at invalid state: {}", __func__);
+    return CAPIStatus::CAPIInvalidPhase;
+  }
+  copyHeaderMapToGo(*headers, go_strs, go_buf);
+  return CAPIStatus::CAPIOK;
+}
+
+// It won't take affect immidiately while it's invoked from a Go thread, instead, it will post a
+// callback to run in the envoy worker thread.
+CAPIStatus TcpUpstream::setRespHeader(ProcessorState& state, absl::string_view key, absl::string_view value,
```

Comment:
we need to check `already_send_resp_headers_` here, and it worth a warning error log when set resp header with `already_send_resp_headers_ = true`.